### PR TITLE
Fixes for diagnostics script with namespace

### PR DESCRIPTION
**Pull request by anders-elastisys**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [x] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change 
- [ ] kind/dev-change   
- [ ] kind/security     
- [ ] [kind/adr]()      

### What does this PR do / why do we need this PR?

Noticed that running `diagnostics` with a `namespace` exited if no logs were found due to the exit code of `grep` and the script setting `set -euo pipefail`. This PR fixes this, and also adds a test for running `diagnostics namespace`.
Also added some additional patterns to `grep` logs for and simplified to get all configmaps in the namespace since we are already getting all pods in the namespace.

- Fixes #

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
  - [ ] The change upgrades CRDs
  - [ ] The change updates the config *and* the schema
- Documentation checks:
  - [ ] The [public documentation](https://github.com/elastisys/compliantkubernetes) required no updates
  - [ ] The [public documentation](https://github.com/elastisys/compliantkubernetes) required an update - [link to change](set-me\)
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [x] The bug fix is covered by regression tests

### CODE REVIEW

**Review by AlbinB97 — CHANGES_REQUESTED** [review]
image
I tested it, and while normal diagnostics script seems to work fine, I get this error when trying to specify a namespace. Maybe I am doing it wrong? But the script says that `-n` should be a valid flag.

**Comment by salehsedghpour** [comment]
I tested it, and while normal diagnostics script seems to work fine, I get this error when trying to specify a namespace. Maybe I am doing it wrong? But the script says that `-n` should be a valid flag.


I checked, to me it worked with using the keyword `namespace` instead of `-n`:
```bash
./bin/ck8s diagnostics wc namespace default
```

So we can either support `-n` or update the hint. I prefer the first one, as it aligns more with our experience with working with namespaces!

File: bin/diagnostics.bash [hunk 1]
```diff
@@ -299,36 +299,18 @@ run_diagnostics_namespaced() {
...
+    "${here}/ops.bash" kubectl "${cluster}" get configmaps -n "${namespace}" -o yaml
     printf '%*s\n' "${COLUMNS:-$(tput cols)}" '' | tr ' ' -
 
     # -- Logs --
     printf '%*s\n' "${COLUMNS:-$(tput cols)}" '' | tr ' ' -
-    echo "Fetching Logs <logs>"
-    pods=$("${here}/ops.bash" kubectl "${cluster}" get pods -n "${namespace}" -o yaml | yq4 '.items[] | .metadata.name')
-    readarray pods_arr < <(echo "$pods" | yq4 e -o=j -I=0 '.[]')
-
+    echo "Fetching error Logs <logs>"
```
Comment (Zash) [inline bin/diagnostics.bash:307]: Consistent capitalization would be good.

File: bin/diagnostics.bash [hunk 2]
```diff
@@ -299,36 +299,18 @@ run_diagnostics_namespaced() {
...
-    cfgs=($(echo "${cfgs[@]}" | tr [:space:] '\n' | awk '!a[$0]++'))
-
-    for cfg in "${cfgs[@]}"; do
-        if [ "$cfg" == "null" ]; then
-            :
-        else
-            configmap=$(sed -e 's/^"//' -e 's/"$//' <<<"$cfg")
-            "${here}/ops.bash" kubectl "${cluster}" get configmap -n "${namespace}" "$configmap" -o yaml
-        fi
-    done
```
Comment (Zash) [inline bin/diagnostics.bash:320]: Am I reading this correctly that it selects `configmaps` in active use?

Is that distinction not needed? If not then less code is better! 😄

**Review by Zash — COMMENTED** [review]
A question and a nit.

Great to see comments explaining the why's :+1:

**Comment by anders-elastisys** [comment]
@salehsedghpour @AlbinB97 PTAL, I changed it to only show the `--help` flag in the `ck8s` usage, since we also support other commands and flags not visible there, and adding them would make it to cluttered. https://github.com/elastisys/compliantkubernetes-apps/pull/2361/commits/15bd0f19d4ebe503b3e148cdc8eee6d573c91546

**Review by salehsedghpour — APPROVED** [review]

**Review by AlbinB97 — APPROVED** [review]
Nice work, thanks! 👍

File: bin/diagnostics.bash (hunk 2 above)
Comment (anders-elastisys) [reply bin/diagnostics.bash:320]: It selected configmaps in active use for pods that are not `Ready`, which can have its uses, but I feel like we might be missing important information by not including ConfigMaps making it more difficult for whoever is going to look at the generated diagnostics to debug an issue. But I see merits with both approaches, as this change does generate more diagnostics, which might also be difficult to debug.

**Review by Elias-elastisys — APPROVED** [review]

File: bin/diagnostics.bash (hunk 2 above)
Comment (Zash) [reply bin/diagnostics.bash:320]: Filtering locally may be faster than the round trip of asking for more information, so makes sense to err on the side of verbosity.

**Review by Zash — APPROVED** [review]

**Pull request merged**